diff -up perl-5.10.1/lib/CPAN/FirstTime.pm.eee perl-5.10.1/lib/CPAN/FirstTime.pm
--- perl-5.10.1/lib/CPAN/FirstTime.pm.eee	2010-06-08 14:00:05.331511360 +0200
+++ perl-5.10.1/lib/CPAN/FirstTime.pm	2010-06-08 11:13:54.758505838 +0200
@@ -952,7 +952,7 @@ sub init {
     my_prompt_loop(prefer_installer => 'MB', $matcher, 'MB|EUMM|RAND');
 
     if (!$matcher or 'makepl_arg make_arg' =~ /$matcher/) {
-        my_dflt_prompt(makepl_arg => "", $matcher);
+        my_dflt_prompt(makepl_arg => "INSTALLDIRS=site", $matcher);
         my_dflt_prompt(make_arg => "", $matcher);
         if ( $CPAN::Config->{makepl_arg} =~ /LIBS=|INC=/ ) {
             $CPAN::Frontend->mywarn( 
@@ -974,7 +974,7 @@ sub init {
     my_dflt_prompt(make_install_arg => $CPAN::Config->{make_arg} || "",
                    $matcher);
 
-    my_dflt_prompt(mbuildpl_arg => "", $matcher);
+    my_dflt_prompt(mbuildpl_arg => "--installdirs site", $matcher);
     my_dflt_prompt(mbuild_arg => "", $matcher);
 
     if (exists $CPAN::HandleConfig::keys{mbuild_install_build_command}
