--- perl-5.10.1/utils/perlbug.PL.fedora	2009-08-12 20:49:24.000000000 +0200
+++ perl-5.10.1/utils/perlbug.PL	2009-11-18 15:56:15.000000000 +0100
@@ -27,8 +27,6 @@ open OUT, ">$file" or die "Can't create 
 open PATCH_LEVEL, "<" . catfile(updir, "patchlevel.h")
     or die "Can't open patchlevel.h: $!";
 
-my $patchlevel_date = (stat PATCH_LEVEL)[9];
-
 while (<PATCH_LEVEL>) {
     last if $_ =~ /^\s*static\s+(?:const\s+)?char.*?local_patches\[\]\s*=\s*{\s*$/;
 }
@@ -71,9 +69,8 @@ $Config{startperl}
     eval 'exec $Config{perlpath} -S \$0 \${1+"\$@"}'
 	if \$running_under_some_shell;
 
-my \$config_tag1 = '$extract_version - $Config{cf_time}';
+my \$config_tag1 = '$extract_version';
 
-my \$patchlevel_date = $patchlevel_date;
 my \$patch_tags = '$patch_tags';
 my \@patches = (
     $patch_desc
@@ -333,17 +330,6 @@ sub Init {
     $ok = '';
     if ($::opt_o) {
 	if ($::opt_o eq 'k' or $::opt_o eq 'kay') {
-	    my $age = time - $patchlevel_date;
-	    if ($::opt_o eq 'k' and $age > 60 * 24 * 60 * 60 ) {
-		my $date = localtime $patchlevel_date;
-		print <<"EOF";
-"perlbug -ok" and "perlbug -nok" do not report on Perl versions which
-are more than 60 days old.  This Perl version was constructed on
-$date.  If you really want to report this, use
-"perlbug -okay" or "perlbug -nokay".
-EOF
-		exit();
-	    }
 	    # force these options
 	    unless ($::opt_n) {
 		$::opt_S = 1; # don't prompt for send
@@ -730,8 +716,8 @@ EFF
     print OUT <<EFF;
 ---
 EFF
-    print OUT "This perlbug was built using Perl $config_tag1\n",
-	    "It is being executed now by  Perl $config_tag2.\n\n"
+    print OUT "This perlbug was built using Perl $config_tag1 in the Fedora build system.\n",
+	    "It is being executed now by Perl $config_tag2.\n\n"
 	if $config_tag2 ne $config_tag1;
 
     print OUT <<EOF;
